File: src/0155-Min-Stack/0155.py
class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data = [] 
        self.min_val = float("inf")

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x <= self.min_val:
            self.data.append(self.min_val)
            self.min_val = x
            
        self.data.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if self.data[-1] == self.min_val:
            self.data.pop()
            self.min_val = self.data[-1]
            self.data.pop()
        else:
            self.data.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.data[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min_val

File: src/0155-Min-Stack/0155.cpp
static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class MinStack 
{
public:
    void push(int x) 
    {
	    s1.push(x);
	    if (s2.empty() || x <= getMin())  s2.push(x);	    
    }
    void pop() 
    {
	    if (s1.top() == getMin())  s2.pop();
	    s1.pop();
    }
    int top() 
    {
	    return s1.top();
    }
    int getMin() 
    {
	    return s2.top();
    }
private:
    stack<int> s1;
    stack<int> s2;
};